pkg/sessions/code.go

```diff
@@ -0,0 +1,65 @@
+package sessions
+
+import (
+	"time"
+
+	"github.com/cozy/cozy-stack/pkg/crypto"
+)
+
+// CodeLen is the number of random bytes used for the session codes
+const CodeLen = 16
+
+// A Code is used to transfer a session from the domain with the stack to a
+// subdomain of an application, when the flat subdomains structure is used.
+// The code is valid:
+//   - only once
+//   - for a short time span (1 minute)
+//   - just for one application.
+type Code struct {
+	Value     string
+	SessionID string
+	AppHost   string
+	ExpiresAt int64
+}
+
+// The list of pending codes
+var codes []*Code
```

Comment:
Wouldn't it be possible to use an authenticated message with expiration instead of a random value for each code ?

That would offload the burden to store all the codes on our side (which will become even more tedious at scale where they couldn't be stored in-memory) and make it constant time to check instead of linear to the number of codes we are asked to generate.